### PR TITLE
NER BERT tutorial was added

### PR DESCRIPTION
**Pull request by SergeyAksenovSA**

### CODE REVIEW

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 1]
```diff
@@ -0,0 +1,1802 @@
+{
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Let's rename named_entity_recognition_bert -> named_entity_recognition

Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Let's rename Legal NER BERT.ipynb into named_entity_recognition_huggingface_bert_pipeline.ipynb

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 2]
```diff
@@ -0,0 +1,1802 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "51494c70-acf6-42c3-b7fb-26e400670a2b",
+   "metadata": {},
+   "source": [
+    "The tutorial starts with the assumption that we have partially annotated dataset of texts.\n",
+    "The data is stored on S3 bucket and splitted into two parts: \n",
+    "* $\\textbf{train}$ (~40%) $-$ annotated data for training\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Remove textbf use * instead

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 3]
```diff
@@ -0,0 +1,1802 @@
...
+    "In this tutorial we will go through the following steps:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "81483c83",
+   "metadata": {},
+   "outputs": [],
+   "source": []
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: In this tutorial we will go through the following steps:
And empty cell without steps (you should add them)

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 4]
```diff
@@ -0,0 +1,1802 @@
...
+   "execution_count": 14,
+   "id": "6a6dae51-a536-4190-9c01-d03dc8382fd7",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import boto3\n",
+    "import glob\n",
+    "import os\n",
+    "import pandas as pd\n",
+    "import torch\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: separate with \n

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 5]
```diff
@@ -0,0 +1,1802 @@
...
+    "from collections import Counter,defaultdict\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: space

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 6]
```diff
@@ -0,0 +1,1802 @@
...
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "eeca537e-ba33-4389-a5ba-bbc97b062515",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "! pip install superannotate==4.4.7 #SA SDK installation\n",
+    "! pip install boto3 # install boto3 client"
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: what about hugging face?

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 7]
```diff
@@ -0,0 +1,1802 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "id": "c64b6c4e-f744-4202-9221-4fac99a09e3b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "SA_PROJECT_NAME = \"ADD SUPERANNOTATE PROJECT NAME\""
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Pleas separate words with _ 
Also change it in Rekognition notebook please

Comment (konyrevdmitriy) [reply named_entity_recognition_bert/Legal-NER.ipynb]: Like this ADD_YOUR_TOKEN_HERE

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 8]
```diff
@@ -0,0 +1,1802 @@
...
+       " 'O'}"
+      ]
+     },
+     "execution_count": 21,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "unique_labels"
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Please add at least print. Or it be better to create some pretty output

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 9]
```diff
@@ -0,0 +1,1802 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c763fb0d-44bb-48df-ba7c-b0648c8b3ea0",
+   "metadata": {},
+   "source": [
+    "We will map each label into its id and vice versa for the BERT model."
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: What is vice versa? Are you Rome citizen? =)

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 10]
```diff
@@ -0,0 +1,1802 @@
...
+   "outputs": [],
+   "source": [
+    "def align_label(txt_tokenized,\n",
+    "                entities,\n",
+    "                label2id):\n",
+    "    label_ids = []\n",
+    "    cnt = 0\n",
+    "    for word_idx, (start,end) in zip(txt_tokenized.word_ids(), txt_tokenized['offset_mapping'][0]):\n",
+    "        if word_idx is None:\n",
+    "            label_ids.append(-100)\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: what is -100, some comments please

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 11]
```diff
@@ -0,0 +1,1802 @@
...
+    "            continue\n",
+    "        found_entity = False\n",
+    "        for entity in entities:\n",
+    "            if entity['start'] <= int(start) and entity['end'] >= int(end) and not found_entity:\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Why did you use int()?

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 12]
```diff
@@ -0,0 +1,1802 @@
...
+    "\n",
+    "    return label_ids"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "142f392e-e19a-45c5-a458-d3f9003b78ee",
+   "metadata": {},
+   "source": [
+    "And now we can create class for our dataset."
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: create child class of torch Dataset

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 13]
```diff
@@ -0,0 +1,1802 @@
...
+   "outputs": [],
+   "source": [
+    "class DataSequence(torch.utils.data.Dataset):\n",
+    "\n",
+    "    def __init__(self, texts, entities=None, label2id=None):\n",
+    "        if not entities:\n",
+    "            entities = [[] for t in texts]\n",
+    "        self.texts = [tokenizer(str(i),\n",
+    "                                padding='max_length',\n",
+    "                                max_length = 512,\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: remove spaces

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 14]
```diff
@@ -0,0 +1,1802 @@
...
+   "id": "789fc352",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class DataSequence(torch.utils.data.Dataset):\n",
+    "\n",
+    "    def __init__(self, texts, entities=None, label2id=None):\n",
+    "        if not entities:\n",
+    "            entities = [[] for t in texts]\n",
+    "        self.texts = [tokenizer(str(i),\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Let's do in other way

configured_tokenizer = lambda text: tokenzier(str(text),  padding='max_length', ... )
self.texts = [configured_tokenizer(text) for text in texts]

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 15]
```diff
@@ -0,0 +1,1802 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "id": "d3c19794",
+   "metadata": {},
+   "outputs": [
+    {
+     "ename": "NameError",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Fix this error please

Comment (konyrevdmitriy) [reply named_entity_recognition_bert/Legal-NER.ipynb]: Seems you forgot to import glob

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 16]
```diff
@@ -0,0 +1,1802 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class BertModel(torch.nn.Module):\n",
+    "\n",
+    "    def __init__(self):\n",
+    "\n",
+    "        super(BertModel, self).__init__()\n",
+    "\n",
+    "        self.bert = BertForTokenClassification.from_pretrained('bert-base-cased', num_labels=len(unique_labels))\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Remove extra \n

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 17]
```diff
@@ -0,0 +1,1802 @@
...
+    "\n",
+    "    def forward(self, input_id, mask, label):\n",
+    "\n",
+    "        output = self.bert(input_ids=input_id, attention_mask=mask, labels=label, return_dict=False)\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Remove extra \n

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 18]
```diff
@@ -0,0 +1,1802 @@
...
+    "            train_label = train_label.to(device)\n",
+    "            mask = train_data['attention_mask'].squeeze(1).to(device)\n",
+    "            input_id = train_data['input_ids'].squeeze(1).to(device)\n",
+    "\n",
+    "            optimizer.zero_grad()\n",
+    "            loss, logits = model(input_id, mask, train_label)\n",
+    "\n",
+    "            for i in range(logits.shape[0]):\n",
+    "\n",
+    "                logits_clean = logits[i][train_label[i] != -100]\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: What is -100? Move it into some constant with explanation

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 19]
```diff
@@ -0,0 +1,1802 @@
...
+    "                label_clean = train_label[i][train_label[i] != -100]\n",
+    "\n",
+    "                predictions = logits_clean.argmax(dim=1)\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Looks like you need to add predict method for Bert Model

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 20]
```diff
@@ -0,0 +1,1802 @@
...
+    "    for epoch_num in range(EPOCHS):\n",
+    "\n",
+    "        total_acc_train = 0\n",
+    "        total_loss_train = 0\n",
+    "\n",
+    "        model.train()\n",
+    "\n",
+    "        for train_data, train_label in tqdm(train_dataloader):\n",
+    "            train_label = train_label.to(device)\n",
+    "            mask = train_data['attention_mask'].squeeze(1).to(device)\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Add some comment for each step, please

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 21]
```diff
@@ -0,0 +1,1802 @@
...
+    "            optimizer.step()\n",
+    "\n",
+    "        model.eval()\n",
+    "\n",
+    "        total_acc_val = 0\n",
+    "        total_loss_val = 0\n",
+    "\n",
+    "        for val_data, val_label in val_dataloader:\n",
+    "\n",
+    "            val_label = val_label.to(device)\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Remove extra \n

Add some comment for each step, please

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 22]
```diff
@@ -0,0 +1,1802 @@
...
+     "output_type": "stream",
+     "text": [
+      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
+      "To disable this warning, you can either:\n",
+      "\t- Avoid using `tokenizers` before the fork if possible\n",
+      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
+      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
+      "To disable this warning, you can either:\n",
+      "\t- Avoid using `tokenizers` before the fork if possible\n",
+      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: set the environment variable TOKENIZERS_PARALLELISM=(true | false) please
this warnings are awful

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 23]
```diff
@@ -0,0 +1,1802 @@
...
+     "output_type": "display_data"
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
+      "To disable this warning, you can either:\n",
+      "\t- Avoid using `tokenizers` before the fork if possible\n",
+      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Same

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 24]
```diff
@@ -0,0 +1,1802 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 34,
+   "id": "0b076867",
+   "metadata": {},
+   "outputs": [
+    {
+     "ename": "NameError",
+     "evalue": "name 'evaluation_data' is not defined",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Fix this errors please

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 25]
```diff
@@ -0,0 +1,1802 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 31,
+   "id": "946d1755-af30-40e4-81a5-2c24be94460b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "evaluation_data = evaluate(model, test_texts, test_entities, label2id)"
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Seems you didn't exec this cell

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 26]
```diff
@@ -0,0 +1,1802 @@
...
+    "label_clean = [x[2] for x in evaluation_data]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 158,
+   "id": "cd08dc96",
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Some comments please

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 27]
```diff
@@ -0,0 +1,1802 @@
...
+   "id": "518ec82b",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "              precision    recall  f1-score   support\n",
+      "\n",
+      " CASE_NUMBER       1.00      1.00      1.00       181\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: Change it, looks like something is wrong

File: named_entity_recognition_bert/Legal-NER.ipynb [hunk 28]
```diff
@@ -0,0 +1,1802 @@
...
+     "output_type": "stream",
+     "text": [
+      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
+      "To disable this warning, you can either:\n",
+      "\t- Avoid using `tokenizers` before the fork if possible\n",
+      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
+      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
+      "To disable this warning, you can either:\n",
+      "\t- Avoid using `tokenizers` before the fork if possible\n",
+      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/Legal-NER.ipynb]: set the environment variable TOKENIZERS_PARALLELISM pleaseeeeeeee

**Review by konyrevdmitriy — COMMENTED** [review]
Seems that most of the comments are critical. Please, fix them asap

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 11 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: because it's tensor

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 1 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 2 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 4 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 27 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 5 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 7 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 9 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: ??

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 13 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 6 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 12 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: It is

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 14 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 15 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 16 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 18 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 24 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 25 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 28 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 23 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 22 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/Legal-NER.ipynb (hunk 17 above)
Comment (SergeyAksenovSA) [reply named_entity_recognition_bert/Legal-NER.ipynb]: +

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 29]
```diff
@@ -0,0 +1,1124 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "d65c0bff-ba69-47ac-8b2f-79b711cd84da",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "bucket_name = \"ADD YOUR BUCKET NAME\" # bucket where the data is stored"
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: ADD_YOUR_BUCKET_NAME

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 30]
```diff
@@ -0,0 +1,1124 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "bff0e565",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "token = \"e7514d15a4e0a2bc3824e864bada3aa12183c2aff6ce8910cd14f719946cc6ba797b17ddb4d0b651t=19366\"\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: PUT_YOUR_TOKEN_HERE and comments with link to documentation with getting a token part

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 31]
```diff
@@ -0,0 +1,1124 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "c1dad3c3",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def align_label(txt_tokenized,\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: -> align_label(txt_tokenized, entities, label2id)

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 32]
```diff
@@ -0,0 +1,1124 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "789fc352",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class DataSequence(torch.utils.data.Dataset):\n",
+    "\n",
+    "    def __init__(self, texts, entities=None, label2id=None):\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: remove extra \n

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 33]
```diff
@@ -0,0 +1,1124 @@
...
+    "                                                      return_tensors=\"pt\",\n",
+    "                                                      return_offsets_mapping=True,\n",
+    "                                                      return_length=True)\n",
+    "        self.texts = [configured_tokenizer(text) for text in texts]\n",
+    "        self.labels = [align_label(i, j, label2id) for i,j in zip(self.texts, entities)]\n",
+    "\n",
+    "        \n",
+    "    def __len__(self):\n",
+    "\n",
+    "        return len(self.labels)\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: remove extra \n

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 34]
```diff
@@ -0,0 +1,1124 @@
...
+    "\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: remove extra \n

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 35]
```diff
@@ -0,0 +1,1124 @@
...
+    "    def get_batch_data(self, idx):\n",
+    "\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: remove extra \n

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 36]
```diff
@@ -0,0 +1,1124 @@
...
+    "        return self.texts[idx]\n",
+    "\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: remove extra \n

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 37]
```diff
@@ -0,0 +1,1124 @@
...
+    "    def get_batch_labels(self, idx):\n",
+    "\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: remove extra \n

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 38]
```diff
@@ -0,0 +1,1124 @@
...
+    "        return torch.LongTensor(self.labels[idx])\n",
+    "\n",
+    "    def __getitem__(self, idx):\n",
+    "\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: remove extra \n

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 39]
```diff
@@ -0,0 +1,1124 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "3f54eb86-cf97-4c84-a25b-b54e49d238e7",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "len(train_texts),len(val_texts),len(test_texts)"
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: spaces

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 40]
```diff
@@ -0,0 +1,1124 @@
...
+    "\n",
+    "        for val_data, val_label in val_dataloader:\n",
+    "\n",
+    "            val_label = val_label.to(device)\n",
+    "            mask = val_data['attention_mask'].squeeze(1).to(device)\n",
+    "            input_id = val_data['input_ids'].squeeze(1).to(device)\n",
+    "\n",
+    "            loss, logits = model(input_id, mask, val_label)\n",
+    "\n",
+    "            for i in range(logits.shape[0]):\n",
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: Same part for train and validation move to some method of class
          
           val_label = val_label.to(device)
            mask = val_data['attention_mask'].squeeze(1).to(device)
            input_id = val_data['input_ids'].squeeze(1).to(device)

            loss, logits = model(input_id, mask, val_label)

            for i in range(logits.shape[0]):
                logits_clean = logits[i][val_label[i] != PAD_TOKEN_ID]
                label_clean = val_label[i][val_label[i] != PAD_TOKEN_ID]

                predictions = logits_clean.argmax(dim=1)
                acc = (predictions == label_clean).float().mean()

File: named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb [hunk 41]
```diff
@@ -0,0 +1,1124 @@
...
+    "\n",
+    "    outputs.append(sa_client.upload_annotations(project=f'{SA_PROJECT_NAME}/unlabeled/', annotations=batch))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2a5fa4cc-d03c-41e1-8195-e8f49c0b33b0",
+   "metadata": {},
+   "source": [
+    "Now we can look at unlabeled folder at the SuperAnnotate page and see the predictions made by our model.\n"
```
Comment (konyrevdmitriy) [inline named_entity_recognition_bert/named_entity_recognition_huggingface_bert_pipeline.ipynb]: Please add also screenshot with marked document by ner

**Review by konyrevdmitriy — COMMENTED** [review]
Some more comments

**Pull request merged**